curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt

```diff
@@ -97,29 +105,27 @@ fun View.onAttachedToWindow(onAttachedToWindow: () -> Unit) {
 }
 
 fun <T : Activity> launchWaitingForFocus(activityClass: KClass<T>): ActivityScenario<T> {
-  val waitForFocus = CountDownLatch(1)
+  val activityHasWindowFocus = ArrayBlockingQueue<Boolean>(1)
   val scenario = ActivityScenario.launch(activityClass.java)
   scenario.onActivity { activity ->
     if (activity.hasWindowFocus()) {
-      waitForFocus.countDown()
+      activityHasWindowFocus.put(true)
     } else {
-      val focusListener = object : OnWindowFocusChangedListener {
+      activity.window.onWindowFocusChangedListeners += object : OnWindowFocusChangedListener {
         override fun onWindowFocusChanged(hasFocus: Boolean) {
-          check(hasFocus) {
-            "onWindowFocusChanged() unexpectedly called with hasFocus false"
-          }
           activity.window.onWindowFocusChangedListeners -= this
-          waitForFocus.countDown()
+          activityHasWindowFocus.put(hasFocus)
```

Comment:
I don't understand what this changes. Prior to this we had `check(hasFocus) {` so that main thread would throw immediately, effectively making the test fail is that callback was to happen with the wrong boolean. So effectively we have the same situation where we wait 30 seconds if we don't get any callback, and fail immediately if we get a callback with the wrong value.

Not that this is necessarily wrong, and generally this seems like a useful helper, I just don't quite see if / how this helps :)

Reply:
If it checked first and threw (so it didn't count down the latch afterwards), then the test thread would still be waiting for the latch, effectively making the failing test always take 30 seconds to complete.

This makes it the test fails near instantly if wrong `hasFocus` comes in.

Reply:
>  then the test thread would still be waiting for the latch, effectively making the failing test always take 30 seconds to complete.

Oooooh

Reply:
Across 12 emulators in the new shiny Android Studio test matrix UI (it's so good)

With `activityWindowFocusChanged.put(!hasFocus)`:

![obrazek](https://user-images.githubusercontent.com/4589773/107586355-0aaf6a00-6c00-11eb-9fcf-73ec3240567e.png)

With `TODO()`:

![obrazek](https://user-images.githubusercontent.com/4589773/107586176-bc9a6680-6bff-11eb-9d0b-59fb8b44d8a6.png)

The exception *on the main thread* happens quickly, yes. But the test still has to wait 30 seconds for the latch/queue. If we deliver the callback value from the main thread *to the test thread*, whatever happens next, the test thread and the test finishes right after that.